lib/commands/media-projection.js

```diff
@@ -0,0 +1,270 @@
+    method: method || 'PUT',
+    headers,
+    fileFieldName,
+    formFields,
+  };
+  if (user && pass) {
+    options.auth = {user, pass};
+  }
+  await net.uploadFile(localFile, remotePath, options);
+  return '';
+}
+
+async function verifyMediaProjectionRecordingIsSupported (adb) {
+  const apiLevel = await adb.getApiLevel();
+  if (apiLevel < MIN_API_LEVEL) {
+    throw new Error(`Media projection-based recording is not available on API Level ${apiLevel}. ` +
+      `Minimum required API Level is ${MIN_API_LEVEL}.`);
+  }
+}
+
+
+class MediaProjectionRecorder {
+  constructor (adb) {
+    this.adb = adb;
+  }
+
+  async isRunning () {
+    const stdout = await this.adb.shell(['dumpsys', 'activity', 'services']);
+    return stdout.includes(RECORDING_SERVICE_NAME);
+  }
+
+  async start (opts = {}) {
+    if (await this.isRunning()) {
+      return false;
+    }
+
+    await this.cleanup();
+    const {
+      filename,
+      maxDurationSec,
+      priority,
+      resolution,
+    } = opts;
+    const args = [
+      'am', 'start',
+      '-n', RECORDING_ACTIVITY_NAME,
+      '-a', RECORDING_ACTION_START,
+    ];
+    if (filename) {
+      args.push(
+        '--es', 'filename',
+        _.toLower(filename).endsWith(DEFAULT_EXT) ? filename : `${filename}${DEFAULT_EXT}`
+      );
+    }
+    if (maxDurationSec) {
+      args.push('--es', 'max_duration_sec', `${maxDurationSec}`);
+    }
+    if (priority) {
+      args.push('--es', 'priority', priority);
+    }
+    if (resolution) {
+      args.push('--es', 'resolution', resolution);
+    }
+    await this.adb.shell(args);
+    await new B((resolve, reject) => {
+      setTimeout(async () => {
+        if (!await this.isRunning()) {
+          return reject(new Error(
+            `The media projection recording is not running after ${RECORDING_STARTUP_TIMEOUT_MS}ms. ` +
+            `Please check the logcat output for more details.`
+          ));
+        }
+        resolve();
+      }, RECORDING_STARTUP_TIMEOUT_MS);
+    });
+    return true;
+  }
+
+  async cleanup () {
+    await this.adb.shell([`rm -f ${RECORDINGS_ROOT}/*`]);
+  }
+
+  async pullRecent () {
+    const recordings = await this.adb.ls(RECORDINGS_ROOT, ['-tr']);
+    if (_.isEmpty(recordings)) {
+      return null;
+    }
+
+    const dstPath = path.join(await tempDir.openDir(), recordings[0]);
+    await this.adb.pull(`${RECORDINGS_ROOT}/${recordings[0]}`, dstPath);
+    return dstPath;
+  }
+
+  async stop () {
+    if (!await this.isRunning()) {
+      return false;
+    }
+
+    await this.adb.shell([
+      'am', 'start',
+      '-n', RECORDING_ACTIVITY_NAME,
+      '-a', RECORDING_ACTION_STOP,
+    ]);
+    try {
+      await waitForCondition(async () => !(await this.isRunning()), {
+        waitMs: RECORDING_STOP_TIMEOUT_MS,
+        intervalMs: 500,
+      });
+    } catch (e) {
+      throw new Error(
+        `The current media projection recording cannot stop within ${RECORDING_STOP_TIMEOUT_MS}ms timeout`
```

Comment:
`The attempt to stop the current media projection recording timed out after ${RECORDING_STOP_TIMEOUT_MS}ms`

Reply:
updated